Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,494 @@
+  | ⟨a₁, h₁⟩, ⟨a₂, h₂⟩ => ⟨a₁ ++ a₂, by rw[Array.size_append, ←h₁,←h₂]⟩
+
+
+instance : HAppend (Vector α n) (Vector α m) (Vector α (n + m)) where
+  hAppend := append
+/--
+  `Vector.head` produces the head of a vector
+-/
+abbrev head (v : Vector α (n+1)) :=
+  v.toArray.get ⟨0, v.size_eq.symm ▸ Nat.zero_lt_succ n⟩
+
+
+/--
+  `Vector.cons` is the list cons pattern for vectors
+-/
+
+def cons (x : α) (xs : Vector α n) : Vector α (n + 1) := Nat.add_comm 1 n ▸ (#v[x] ++ xs)
+
+/--Make a vector of size `n` with all cells containing `v`-/
+def mkVector (n : Nat) (v : α) : Vector α n :=
+  ⟨mkArray n v, proof⟩
+  where
+    proof := by
+      simp only [size_mkArray]
+      done
+
+/-- Create empty vector `#v[]`-/
+def mkVector0 : Vector α 0 := empty
+
+/-- Create vector `#v[v₁]` -/
+def mkVector1 (v₁ : α) : Vector α 1 :=
+  mkEmpty 1 |> push v₁
+
+/-- Create vector `#v[v₁, v₂]` -/
+def mkVector2 (v₁ v₂ : α) : Vector α 2 :=
+  mkEmpty 2
+    |> push v₁
+    |> push v₂
+
+/-- Create vector `#v[v₁, v₂, v₃]` -/
+def mkVector3 (v₁ v₂ v₃ : α) : Vector α 3:=
+  mkEmpty 3
+    |> push v₁
+    |> push v₂
+    |> push v₃
+
+
+/-- Create vector `#v[v₁, v₂, v₃, v₄]` -/
+def mkVector4 (v₁ v₂ v₃ v₄ : α) : Vector α 4:=
+  mkEmpty 4
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+
+/-- Create vector `#v[v₁, v₂, v₃, v₄, v₅]` -/
+def mkVector5 (v₁ v₂ v₃ v₄ v₅ : α) : Vector α 5 :=
+  mkEmpty 5
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+
+/-- Create vector `#v[v₁, v₂, v₃, v₄, v₅, v₆]` -/
+def mkVector6 (v₁ v₂ v₃ v₄ v₅ v₆ : α) : Vector α 6 :=
+  mkEmpty 6
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+
+/-- Create vector `#v[v₁, v₂, v₃, v₄, v₅, v₆, v₇]` -/
+def mkVector7 (v₁ v₂ v₃ v₄ v₅ v₆ v₇ : α) : Vector α 7 :=
+  mkEmpty 7
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+    |> push v₇
+
+/-- Create vector `#v[v₁, v₂, v₃, v₄, v₅, v₆, v₇, v₈]` -/
+def mkVector8 (v₁ v₂ v₃ v₄ v₅ v₆ v₇ v₈ : α) : Vector α 8 :=
+  mkEmpty 8
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+    |> push v₇
+    |> push v₈
+
+/-- `extract v start halt` Returns the slice of `v` from indices `start` to `stop` (exclusive).
+  If `start` is greater or equal to `stop`, the result is empty.
+  If `stop` is greater than the size of `v`, the size is used instead.-/
+def extract (v : Vector α n)
+  (start : Nat) (halt : Nat) : Vector α (min halt n - start) :=
+  ⟨Array.extract v.toArray start halt, proof⟩
+  where
+    proof := by
+      simp only [size_extract]
+      have size_eq : n = v.toArray.size := by
+        exact v.size_eq.symm
+      rw [←size_eq]
+      done
+
+
+/-- Eliminator for `Vector α n`-/
+@[elab_as_elim]
+def elim {C : ∀ {n}, Vector α n → Sort u}
+    (H : ∀ a : Array α, C ⟨a, rfl⟩) {n : Nat} : ∀ v : Vector α n, C v
```

Comment:
Custom eliminators are often useful but I'm not immediately seeing the use case for this one. Any example uses in Mathlib?

Reply:
No idea about this. But I just noticed that the signature is wrong.

Reply:
But we have so far not written many proofs with vectors purely using Vector's API. So I don't want to rule out the possibility

Reply:
I've played around with it a bit and I see the utility now. I recommend replacing with these two custom eliminators and moving closer to the top of the file.
```lean
/-- Custom eliminator for `Vector α n` -/
@[elab_as_elim]
def elimAsArray {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : Array α, motive ⟨a, rfl⟩) :
    {n : Nat} → (v : Vector α n) → motive v
  | _, ⟨a, rfl⟩ => mk a

/-- Custom eliminator for `Vector α n` -/
@[elab_as_elim]
def elimAsList {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : List α, motive ⟨⟨a⟩, rfl⟩) :
    {n : Nat} → (v : Vector α n) → motive v
  | _, ⟨⟨a⟩, rfl⟩ => mk a
```

Reply:
Done.